Reintroduce *BSD support

smortex opened the pull request:
Hello !

#220 suggests that \*BSD support is broken, here is a fix!

This was tested on FreeBSD 10.1-STABLE, but does not touch to *low-level* things so if `rb-kqueue` works as expected on other BSD variants, it's expected to be fine there too :smile: .

~~~
Finished in 6 minutes 22 seconds (files took 2.57 seconds to load)
349 examples, 0 failures
~~~

I removed startup warnings because it cluttered a lot the output of `rspec`, but I did not changed the `README.md` file not removed the warning from the Gemfile as I don't know what you would prefer:

* Reintegrate \*BSD as supported platforms;
* Keep some *here be dragons* warning;
* Drop the BSD stuff (I don't hope so!).

Just tell me what you would like and I'll add a few commits to the pull request accordingly.

Thanks!

e2 commented:
First, thanks very much for the PR! 

Overall, I'm glad to accept *BSD patches - since I can't justify working on those myself (in favor of other issues) - so don't hesitate to send PRs, ask, etc.

The areas I'm most worried about are:
- does it work on VMs with BSDs (last time I checked it didn't)
- is there are particular version of Celluloid it needs to be locked to to work (is there a released version that does - and also on VMs?)
- is it possible to have a CI setup somewhere (I guess Travis won't have BSD support for a long while, see: https://github.com/travis-ci/travis-ci/issues/1818)
- also, make sure listening with multiple directories works (you'll need to check manually, because there are no tests for that AFAIR)

Ideally, there would be some unit tests with mocks for rb-kqueue (see darwin_spec.rb for ideas), so I can add include potentially upcoming encoding-related fixes for BSD also.

Also, updates to the README

Thanks again!

e2 commented:
A few notes:

1) it's best to think that listen "watches directories and reports changes to files" - so you can ignore watching files specifically if you can watch a directory and get file events
2) there is "special editor support" in listen, so you may want to check e.g. 'sed -i' commands and make sure they are reported as "modified" operations (and not "added") 
3) you can squash the commits - no problem there
4) manually check if watching multiple directories generates the correct events (as I've mentioned above) - since there are no tests for this

It looks good to me, so let me know when I can merge this.

smortex commented:
Hi @e2,

I'll try to answer your questions

## Does it works on a VM

I don't know: I don't believe in VMs, only OS on bare metal :smiling_imp:.

## Did I used a particular version of *insert-name-here*

Not at all: I just wanted to run Jekyll on FreeBSD, got the warning from *listen* that I had to use some particular gems in my Gemfile to avoid polling, give it a try but it failed because it referenced repositories that do not exist anymore, so I tried to fix *listen* directly and this lead me here…

Here is my `Gemfile`:

~~~ruby
source 'https://rubygems.org'

gem 'i18n'
gem 'jekyll'
gem 'jekyll-haml'
gem 'listen', github: 'sante-link/listen', branch: 'bsd'

require 'rbconfig'
if RbConfig::CONFIG['target_os'] =~ /bsd|dragonfly/
  gem 'rb-kqueue', '>= 0.2', require: false
end
~~~

For the record, here is the output of `bundle show`:

~~~
Gems included by the bundle:
  * blankslate (2.1.2.4)
  * bundler (1.7.11)
  * celluloid (0.16.0)
  * classifier-reborn (2.0.3)
  * coffee-script (2.3.0)
  * coffee-script-source (1.8.0)
  * colorator (0.1)
  * execjs (2.2.2)
  * fast-stemmer (1.0.2)
  * ffi (1.9.6)
  * haml (4.0.6)
  * hitimes (1.2.2)
  * i18n (0.7.0)
  * jekyll (2.5.3)
  * jekyll-coffeescript (1.0.1)
  * jekyll-gist (1.1.0)
  * jekyll-haml (0.1.1)
  * jekyll-paginate (1.1.0)
  * jekyll-sass-converter (1.3.0)
  * jekyll-watch (1.2.0)
  * kramdown (1.5.0)
  * liquid (2.6.1)
  * listen (2.8.4 2815e74)
  * mercenary (0.3.5)
  * parslet (1.5.0)
  * posix-spawn (0.3.9)
  * pygments.rb (0.6.0)
  * rb-fsevent (0.9.4)
  * rb-inotify (0.9.5)
  * rb-kqueue (0.2.3)
  * redcarpet (3.2.2)
  * safe_yaml (1.0.4)
  * sass (3.4.9)
  * tilt (2.0.1)
  * timers (4.0.1)
  * toml (0.1.2)
  * yajl-ruby (1.1.0)
~~~

And it does the job :smile: !

## BSD Continuous Integration system

No idea, sorry.  My development machine runs FreeBSD but our servers run Linux, so I can't afford a CI system.

## More testing

Here is a bunch of suggested tests from `irb` (absolute path to local directory replaced by `/...` for privacy).  These results look consistent, please tell me if something comes to your eyes.

~~~
$ bundle exec irb
irb(main):001:0> require 'listen'
=> true
irb(main):002:0> listener = Listen.to('_i18n', '_people', '_layouts') do |modified, added, removed|
irb(main):003:1*   puts "modified absolute path: #{modified}"
irb(main):004:1>   puts "added absolute path: #{added}"
irb(main):005:1>   puts "removed absolute path: #{removed}"
irb(main):006:1> end
irb(main):007:0> listener.start
~~~

File creation with `trouch _layouts/x` reports:

~~~
modified absolute path: []
added absolute path: ["/.../_layouts/x"]
removed absolute path: []
~~~

File removal with `rm _layouts/x` reports:

~~~
modified absolute path: []
added absolute path: []
removed absolute path: ["/.../_layouts/x"]
~~~

Rename file with `mv _i18n/en.yml _i18n/us.yml` reports:

~~~
modified absolute path: []
added absolute path: ["/.../_i18n/us.yml"]
removed absolute path: ["/.../_i18n/en.yml"]
~~~

Change in a file with `sed -i "" "s/Foo/Bar/" _i18n/us/yml` reports:

~~~
modified absolute path: ["/.../_i18n/us.yml"]
added absolute path: []
removed absolute path: []
~~~

e2 commented:
> I don't believe in VMs, only OS on bare metal :smiling_imp:.

I understand completely - I use Linux on bare metal (I wouldn't mind BSD if it wasn't so "lonely"), and unfortunately my "dev setup" takes 2-3 days to complete, so I can't afford "switching" to a new OS. Also, I use VMs as "snapshots" and as "chroot jails" (when I'm not using docker).

The results look good. You can add changes to the README, etc.

The last commit shows an error - because a directory is reported in the results (should be files only).

If that works, just rebase/squash and let me know.

smortex commented:
The last commit *fix* this error ;-)

I'll fix the README and Gemfile to remove warnings !

e2 commented:
LGTM. Can I merge this?

smortex commented:
I guess so :smile:.

Thanks!

e2 commented:
Awesome job, thanks! 

(I really appreciate it, since FreeBSD was what got me hooked on UNIX 14 years ago!)

webhat commented:
:+1: *BSD

e2 commented:
Released as Listen 2.8.5
